README
EasyMineLauncher
----------------

An easy and lightweight launcher for Minecraft which takes some of the hassle
away.

Due to the way Minecraft is launched beginning with the 1.6 snapshots and the 1.6 release it is not possible to launch these versions with EasyMineLauncher.


What it does:
-------------

 * Removes the need to login if you just want to play offline or on
   a private server
 * There's no check for an update, you're directly catapulted to the main menu
   or connected to a server
 * Additional parameters (f.e. start as maximized window or loading of
   certain versions from different directories)


What it does not:
-----------------

 * Remove the need to authenticate at servers if they demand it
 * Loading and applying updates
 * Lets you play Minecraft for free, you'll still need to get an account and
   download the binaries


How does it work?
-----------------

Basically this is just mimicking the behavior of the official launcher,
except that it removes the online authentication and the update process.
I figured most of that out with Trial & Error, that's why it's looking
a little bit crude and has debugging output all over the place.


Compatibility
-------------

This launcher is compatible with the following versions:

 * Indev
 * Infdev
 * Alpha
 * Beta
 * 1.0 - 1.5

This launcher does not work with the 1.6 snapshots and release or any later version.


Where to get it?
----------------

Ready-to-use jars are available from [the Download-Section](http://easyminelauncher.bonsaimind.org/).


What can I do with it?
----------------------

Awesome stuff?

Here's the output from `--help`:

    Usage: EasyMineLauncher.jar [OPTION]
    Launch Minecraft directly.
    
    --help                   Prints this help.
    --version                Prints the version.


    --jar-dir=DIRECTORY      Set the directory for the jar files.
    --jar=MINECRAFT.JAR      Set the path to the minecraft.jar.
                             Either specify jar-dir or this.
    --lwjgl-dir=DIRECTORY    Set the directory for the jar files
                             of lwjgl (lwjgl.jar, lwjgl_util.jar, and jinput.jar)
    --password=PASSWORD      Set the password.
    --mppass=MPPASS          Same as --password=PASSWORD (backwards compatibility)
    --native-dir=DIRECTORY   Set the directory for the native files of lwjgl.
    --additional-jar=JAR     Load the specified jars.
                             This might be needed by some mods.
                             Specify multiple times or list separated with ','.
    --parent-dir=DIRECTORY   Set the parent directory. This effectively changes
                             the location of the .minecraft folder.
    --applet=APPLET          The (full) name of the applet to load. Normally this
                             points to the Minecraft-Applet.
    --blend-with=JAR         Blend the (found) minecraft.jar with JAR.
                             That means that the content of JAR is copied into
                             a temporary jar and then is supplemented with the
                             missing files from minecraft.jar.
                             Please be aware that this means that two archives are
                             merged together at every start, that takes on my SSD
                             a whooping 1.2 seconds.
    --blend-jar-name=NAME    The NAME of the blended jar, defaults to
                             minecraft_blended.jar.
    --blend-keep-manifest    Copy the manifest from the given JAR to blend...that
                             might not work.
    --port=PORT              Set the port of the server, if not set it will revert
                             to 25565.
    --texturepack=FILE       Set the texturepack to use, this takes only
                             the filename (including extension).
                             Use 'Default' for default.
    --server=SERVER          Set the address of the server which directly
                             to connect to.
    --authenticate           Use the given username and password to authenticate
                             at the Mojang server. This will override the given
                             session id (if given) and the username with that
                             returned from Mojang (case corrected).
    --authentication-failure=Set the behavior if the authentication fails,
                             following values are possible:
      SILENT_CONTINUE        Just continue, there will only be a notification
                             on the error output.
      ALERT_CONTINUE         Display a dialog with the reason why it failed,
                             but still continue on.
      SILENT_BREAK           Stop execution and exit, there will only be
                             a notification on the error output.
      ALERT_BREAK            Stop execution and exit, and also display a dialog
                             with the reason why authentication failed.
                             This is the default behavior.
    --keep-alive-tick=TICK   Set the interval (in seconds) in which the
                             authentication at the Mojang server is renewed.
                             Default is 300 seconds.
    --session-id=SESSIONID   Set the session id.
    --launcher-version=VER   Report this launcher version when authenticating.
                             Default is "381".
    --auth-address=ADDRESS   Authenticate at this address instead of
                             the Mojang server.
    --options-file=FILE      Copy the options.txt from this location.
                             Please be aware that this overwrites the options.txt
                             in the Minecraft directory. This will also be done
                             before --texturepack and --set-option.
    --set-option=NAME:VALUE  Set this option in the options.txt file.
    --username=USERNAME      Set the username to user.
    --use-lastlogin          Read username and password from the lastlogin file
                             within the parentdir. Overrides provided username and
                             passwords.
    --save-lastlogin         Writes back the username and password into
                             the lastlogin file, but only after successfull
                             authentication (therefor --authenticate must be
                             provided, too).
    --keep-username          Do not override the username with that returned from
                             Mojang. Only applies when --authenticate is provided.
    --demo                   Trigger Demo-Mode. This might break other stuff.


    --title=TITLE            Replace the window title.
    --height=HEIGHT          The width of the window.
    --width=WIDTH            The height of the window.
    --x=X                    The x-location of the window.
    --y=Y                    The y-location of the window.
    --maximized              Maximize the window.
    --no-frame               Remove the border of the window.
    --always-on-top          Make the window stay above all others.
    --fullscreen             Makes the window the same size as the whole desktop.
                             This is basically shorthand for
                               --width=SCREENWIDTH
                               --height=SCREENHEIGHT
                               --x=0
                               --y=0
                               --no-frame
                               --always-on-top
                             This might yield odd results in multi-monitor
                             environments.
    --opacity=OPACITY        Sets the opacity of the window, were 1 is fully
                             visible, 0 is invisble and 0.5 is half visible.
                             This might or might not work depending on your WM and
                             or compositor. This also forces --no-frame.
    --dump                   Chickens out before doing anything and prints all
                             of the configuration values.
    --no-exit                Does not exit on failure (if Minecraft could not be
                             loaded), only returns.

You can also keep different versions around and launch them with ease,
f.e. with such a directory/file-layout:

    bin
     +- Beta 1.6.6.jar (renamed minecraft.jar)
     +- Beta 1.7.3.jar
     +- Beta 1.8.2.jar
     +- lwjgl
         +- lwjgl jars go here
     +- natives
         +- lwjgl natives go here

With the following command you can launch the versions:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --jar=./bin/Beta \1.6.6.jar --lwjgl-dir=./bin/lwjgl/ --native-dir=./bin/natives/ --username=Terminator


Notes on the parameters
-----------------------

### --additional-jar=JAR

This will allow you to load other jars into memory. Please be aware that this
does nothing but load the jars so that they're available via reflection. There
are some mods which need external jars or the official launcher...you can
specify them with this.

You can specify this either multiple times:

      --additional-jar=./first.jar --additional-jar=./second.jar --additional-jar=./third.jar

Or you can specify a list, whatever suits your needs:

      --additional-jar=./first.jar,./second.jar,./third.jar

There's also the possibility that this might not work, drop an Issue with
the mod and I'll have a look.

### --parent-dir=PARENTDIR

This will allow you to change the location of the minecraft directory which
is normally located in `~/`. Please be aware that this will only change
the location of the `.minecraft` directory, f.e. `/mnt/data/Minecraft/Data/`
becomes `/mnt/data/Minecraft/Data/.minecraft/`.

### --blend-with=JAR

Please be aware that this option does merge two archives together. So depending
in your machine, this may take between 1 and 10 seconds *every start*.


Great...so how do I compile this?
---------------------------------

### The easy way

Download the ready-to-use jar from [the Download-Section](http://easyminelauncher.bonsaimind.org/).

### The normal way

Run `ant`.

### The hard way

 1. Create a manifest file `MANIFEST.MF` in the base directory with the following content:

        Manifest-Version: 1.0
        Main-Class: org.bonsaimind.easyminelauncher.Main

 2. Execute the following commands:

        $ mkdir build dist
        $ javac -d build src/org/bonsaimind/easyminelauncher/*
        $ jar -cvfm dist/EasyMineLauncher.jar MANIFEST.MF -C build build/org/bonsaimind/easyminelauncher/*

 3. Find the ready-to-use jar in `dist`.


Examples
--------

Launch whatever is installed (same as hitting "Offline Mode" in the official launcher):

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar

Launch the current version of Minecraft with a certain username:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --username=MyAwesomeName

Launch the current version of Minecraft with an updated lwjgl version which resides in a different directory:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --lwjgl-dir=/home/user/updatedLWJGL/ --native-dir=/home/user/updatedLWJGL/native/ --username=MyNotSoAwesomeName --maximized
    
Launch the current version of Minecraft and directly connect to a server:

    java -Xmx1536M -Xms1536M -jar EasyMineLauncher.jar --username=MyAwesomeName --server=myprivate.server.com

Launch specific jar, with different lwjgl directory, connect to a server on a different port, maximize the window and rename it:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --jar=./bin/Beta\ 1.7.3.jar/ --lwjgl-dir=./bin/lwjgl/ --native-dir=./bin/natives/ --username=Octiface --server=awesome.server.com --port=666 --maximized --title=UberGame
    
A simple portable launcher:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --jar-dir=./bin/ --parent-dir=./

Authenticate at Mojang:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --username=YourOfficialName --password=shutup --authenticate

Or maybe like this:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --use-lastlogin --authenticate

Some more authentication options:

    java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar --username=YourUsername --password=shooooootu --authenticate --keep-alive=240 --authentication-failure=ALERT_CONTINUE


Not so fast...
--------------

...okay, step by step.

How do make your Minecraft portable, if you already have a `.minecraft`
directory, skip to step 4:

 1. Start the official launcher, download the binaries
 2. Launch Minecraft one time to create the data directory and necessary files
 3. Exit Minecraft
 4. Copy the `~/.minecraft` directory to `/media/yourUsbStick/`
 5. Rename it to something more meaningful
 6. Create a sript with the following contents:
 
        #!/bin/sh
        # Launcher.sh, /media/yourUsbStick/MinecraftPortable/
        
        java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar \
          --jar-dir=./bin/ \
          --parent-dir=./ \
          --username=YourOfflineUsername
 
 7. Make the script executable, done.
 
Okay, now to something more sophisticated, let's extract and split
the `.minecraft` directory so that we can have multiple versions
with ease (assuming you already have different versions and a `.minecraft` dir):

 1. Create a directory as target, e.g. `~/Minecraft/`
 2. Create the following sub-directories: `bin`, `data`, `lwjgl` and `natives`
 3. Copy `~/.minecraft/bin/minecraft.jar` to `~/Minecraft/bin/VERSION.jar`
 4. Copy `~/.minecraft/bin/*.jar` to `~/Minecraft/lwjgl/*.jar`
 5. Copy `~/.minecraft/bin/natives/*.so` to `~/Minecraft/natives/*.so`
 6. Copy everything from `~/.minecraft/` to `~/Minecraft/.minecraft/` except the `bin` directory
 7. Copy the `EasyMineLauncher.jar` into `~/Minecraft/`
 8. Create the `Launcher.sh` script and make it executable:
 
        #!/bin/sh
        # Launcher.sh, ~/Minecraft/
    
        java -Xmx1024M -Xms1024M -jar EasyMineLauncher.jar \
          --jar=./bin/$1.jar \
          --lwjgl-dir=./lwjgl/ \
          --native-dir=./natives/ \
          --parent-dir=./data/ \
          --username=YourOfflineUsername \
          --title=Minecraft\ $1
    
 9. Create a version specific script and make it executable:
 
        #!/bin/sh
        # VERSION.sh, ~/Minecraft/
        
        ./Launcher.sh VERSION
    
 10. Repeat steps 3 and 9 to add further versions.


Problems? Let's see...
----------------------

### ClassNotFoundException

If you're using a mod and you're getting a `ClassNotFoundException: net.minecraft.Launcher` error,
you need to specify the original Minecraft-Launcher with the `--additional-jar` option.

### Dude, my achievements are gone!

Oh yeah, should have told you before, somehow version hopping seems to mess up the achievements, so you might lose all of them.


CHANGES
1.0
---

 * Replaced internals with MinecraftMiddleKnife
 * As a side effect, it is now easily possible to replace specific lwjgl jars.


0.16.1
------

 * Only providing --parent-dir should now work better.
 * Recompiled jar to be compatible with Java 1.5 and up.


0.16
----

 * Fixed a misleading exception message
 * Added options:
   * --auth-address
   * --opacity
   * --dump
   * --no-exit
   * --applet
   * --blend-with, --blend-jar-name, --blend-keep-manifest


0.15
----

 * Fixed --authentication-failure
 * Added options:
   * --use-lastlogin
   * --save-lastlogin


0.14
----

 * Fixed authentication with e-mail address, that should be replaced with
   the username.
 * Added option:
   * --keep-username


0.13
----

 * Added Mojang Authentication with the following options:
   * --authenticate
   * --authentication-failure
   * --keep-alive
   * --launcher-version


0.12
----

 * Added more options:
   * --options-file
 * Fixed:
   * --set-options works now without --texturepack


0.11
----

Fixed compatibility with Forge ModLoader...at least I hope so.


0.10
----

 * Added more options:
   * --demo

This version should add compatibility with Forge ModLoader.


0.9
---

The following parameters are now correctly recognized:
 * --width
 * --height
 * --x
 * --y


0.8
---

 * Added even more options:
   * --x
   * --y
 * Added build-script for ant
 * Fixed hanging/not closable window
 * Fixed possible location issue with --fullscreen

I've tested compatibility with versions through Beta, Alpha back to Indev and
was able to launch all versions I got a hold off.

Also, if you don't specify any parameters with this version, it will fall back
to your Default-Minecraft install and launch that.


0.7
---

 * Added more options:
   * --fullscreen
   * --always-on-top
   * --additional-jar

Please note that the --additional-jar feature might fail for you, that depends
on the mod which tries to load additional classes and the method it uses.
It *should* work for most cases.


0.6
---

 * Added more options:
   * --set-option=NAME:VALUE
   * --no-frame
   * --session-id=SESSIONID
   * --mppass=MPPASS


0.5
---

 * Added possibility to set the texturepack:
   * texturepack


0.4
---

 * Added possibility to change the location of .minecraft:
   * parent-dir


0.3
---

 * Added additional parameters:
   * jar
   * lwjgl-dir
   * native-dir

   
0.2
---

 * Added possibility to directly connect to server.


0.1
---

 * Initial Release


src/org/bonsaimind/easyminelauncher/Parameters.java
/*
 * Copyright 2014 Robert 'Bobby' Zenz. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of
 * conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list
 * of conditions and the following disclaimer in the documentation and/or other materials
 * provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY Robert 'Bobby' Zenz ''AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Robert 'Bobby' Zenz OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of Robert 'Bobby' Zenz.
 */
package org.bonsaimind.easyminelauncher;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import org.bonsaimind.minecraftmiddleknife.pre16.Authenticator;

/**
 * A simple helper class to deal with the commandline arguments.
 */
public class Parameters {
	
	private List<String> additionalJars = new ArrayList<String>();
	private boolean alwaysOnTop = false;
	private String appletToLoad = "net.minecraft.client.MinecraftApplet";
	private boolean authenticate = false;
	private String authenticationAddress = Authenticator.MOJANG_SERVER;
	private AuthenticationFailureBehavior authenticationFailureBehavior = AuthenticationFailureBehavior.ALERT_BREAK;
	private String blendJarName = "minecraft_blended.jar";
	private boolean blendKeepManifest = false;
	private List<String> blendWith = new ArrayList<String>();
	private boolean demo = false;
	private boolean exitRequested = false;
	private boolean fullscreen = false;
	private int height = 600;
	private String jar = "";
	private String jarDir = "";
	private int keepAliveTick = 300;
	private boolean keepUsername = false;
	private String launcherVersion = Authenticator.DEFAULT_LAUNCHER_VERSION;
	private String lwjglDir = "";
	private boolean maximized = false;
	private String nativeDir = "";
	private boolean noExit = false;
	private boolean noFrame = false;
	private float opacity = 1f;
	private List<String> options = new ArrayList<String>();
	private String optionsFileFrom = "";
	private String parentDir = System.getProperty("user.home");
	private String password = "";
	private String port = "25565";
	private boolean printDump = false;
	private boolean printHelp = false;
	private boolean printVersion = false;
	private boolean saveLastLogin = false;
	private String server = null;
	private String sessionId = "0";
	private String texturepack = "";
	private String title = "Minecraft";
	private boolean useLastLogin = false;
	private String username = "Username";
	private int width = 800;
	private int x = -1;
	private int y = -1;
	
	public Parameters(String[] arguments) {
		for (String arg : arguments) {
			if (arg.startsWith("--jar-dir=")) {
				jarDir = arg.substring(10);
			} else if (arg.startsWith("--jar=")) {
				jar = arg.substring(6);
			} else if (arg.startsWith("--lwjgl-dir=")) {
				lwjglDir = arg.substring(12);
			} else if (arg.startsWith("--mppass=")) {
				password = arg.substring(9);
			} else if (arg.startsWith("--password=")) {
				password = arg.substring(11);
			} else if (arg.startsWith("--native-dir=")) {
				nativeDir = arg.substring(13);
			} else if (arg.startsWith("--additional-jar=")) {
				for (String additionalJar : arg.substring(17).split(",")) {
					if (additionalJar.length() > 0) {
						additionalJars.add(additionalJar);
					}
				}
			} else if (arg.equals("--no-frame")) {
				noFrame = true;
			} else if (arg.startsWith("--parent-dir=")) {
				parentDir = arg.substring(13);
			} else if (arg.startsWith("--applet=")) {
				appletToLoad = arg.substring(9);
			} else if (arg.startsWith("--blend-with=")) {
				for (String blendWithJar : arg.substring(13).split(",")) {
					if (blendWithJar.length() > 0) {
						blendWith.add(blendWithJar);
					}
				}
			} else if (arg.startsWith("--blend-jar-name=")) {
				blendJarName = arg.substring(17);
			} else if (arg.equals("--blend-keep-manifest")) {
				blendKeepManifest = true;
			} else if (arg.startsWith("--port=")) {
				port = arg.substring(7);
			} else if (arg.startsWith("--server=")) {
				server = arg.substring(9);
			} else if (arg.equals("--authenticate")) {
				authenticate = true;
			} else if (arg.startsWith("--authentication-failure=")) {
				authenticationFailureBehavior = AuthenticationFailureBehavior.valueOf(arg.substring(25));
			} else if (arg.startsWith("--keep-alive-tick=")) {
				keepAliveTick = Integer.parseInt(arg.substring(18));
			} else if (arg.startsWith("--session-id=")) {
				sessionId = arg.substring(13);
			} else if (arg.startsWith("--launcher-version=")) {
				launcherVersion = arg.substring(19);
			} else if (arg.startsWith("--auth-address=")) {
				authenticationAddress = arg.substring(15);
			} else if (arg.startsWith("--options-file=")) {
				optionsFileFrom = arg.substring(15);
			} else if (arg.startsWith("--set-option=")) {
				options.add(arg.substring(13));
			} else if (arg.startsWith("--texturepack=")) {
				texturepack = arg.substring(14);
			} else if (arg.startsWith("--title=")) {
				title = arg.substring(8);
			} else if (arg.startsWith("--username=")) {
				username = arg.substring(11);
			} else if (arg.equals("--use-lastlogin")) {
				useLastLogin = true;
			} else if (arg.equals("--save-lastlogin")) {
				saveLastLogin = true;
			} else if (arg.equals("--keep-username")) {
				keepUsername = true;
			} else if (arg.equals("--demo")) {
				demo = true;
			} else if (arg.equals("--version")) {
				printVersion = true;
			} else if (arg.startsWith("--width=")) {
				width = Integer.parseInt(arg.substring(8));
			} else if (arg.startsWith("--height=")) {
				height = Integer.parseInt(arg.substring(9));
			} else if (arg.startsWith("--x=")) {
				x = Integer.parseInt(arg.substring(4));
			} else if (arg.startsWith("--y=")) {
				y = Integer.parseInt(arg.substring(4));
			} else if (arg.equals("--maximized")) {
				maximized = true;
			} else if (arg.equals("--always-on-top")) {
				alwaysOnTop = true;
			} else if (arg.equals("--fullscreen")) {
				fullscreen = true;
			} else if (arg.startsWith("--opacity=")) {
				opacity = Float.parseFloat(arg.substring(10));
			} else if (arg.equals("--dump")) {
				printDump = true;
			} else if (arg.equals("--no-exit")) {
				noExit = true;
			} else if (arg.equals("--help")) {
				printHelp = true;
			} else {
				printHelp = true;
			}
		}
		
		// Check if we were provided with a path, otherwise fall back to the
		// defaults.
		if (jarDir.isEmpty() && jar.isEmpty()) {
			jarDir = new File(new File(parentDir, ".minecraft").toString(), "bin").getAbsolutePath();
			jar = new File(jarDir, "minecraft.jar").getAbsolutePath();
		} else if (jarDir.isEmpty()) {
			jarDir = new File(jar).getParent();
		}
		
		if (lwjglDir.isEmpty()) {
			lwjglDir = jarDir;
		}
		
		if (nativeDir.isEmpty()) {
			nativeDir = new File(jarDir, "natives").getAbsolutePath();
		}
		
		// Required additional jars.
		extendAdditionalJars(new File(lwjglDir, "jinput.jar").getAbsolutePath());
		extendAdditionalJars(new File(lwjglDir, "lwjgl.jar").getAbsolutePath());
		extendAdditionalJars(new File(lwjglDir, "lwjgl_util.jar").getAbsolutePath());
		
		// Sanity checks
		if (height <= 0) {
			height = 600;
		}
		if (width <= 0) {
			width = 800;
		}
	}
	
	public List<String> getAdditionalJars() {
		return additionalJars;
	}
	
	public String getAppletToLoad() {
		return appletToLoad;
	}
	
	public String getAuthenticationAddress() {
		return authenticationAddress;
	}
	
	public AuthenticationFailureBehavior getAuthenticationFailureBehavior() {
		return authenticationFailureBehavior;
	}
	
	public String getBlendJarName() {
		return blendJarName;
	}
	
	public List<String> getBlendWith() {
		return blendWith;
	}
	
	public int getHeight() {
		return height;
	}
	
	public String getJar() {
		return jar;
	}
	
	public String getJarDir() {
		return jarDir;
	}
	
	public int getKeepAliveTick() {
		return keepAliveTick;
	}
	
	public String getLauncherVersion() {
		return launcherVersion;
	}
	
	public String getLwjglDir() {
		return lwjglDir;
	}
	
	public String getNativeDir() {
		return nativeDir;
	}
	
	public float getOpacity() {
		return opacity;
	}
	
	public List<String> getOptions() {
		return options;
	}
	
	public String getOptionsFileFrom() {
		return optionsFileFrom;
	}
	
	public String getParentDir() {
		return parentDir;
	}
	
	public String getPassword() {
		return password;
	}
	
	public String getPort() {
		return port;
	}
	
	public String getServer() {
		return server;
	}
	
	public String getSessionId() {
		return sessionId;
	}
	
	public String getTexturepack() {
		return texturepack;
	}
	
	public String getTitle() {
		return title;
	}
	
	public String getUsername() {
		return username;
	}
	
	public int getWidth() {
		return width;
	}
	
	public int getX() {
		return x;
	}
	
	public int getY() {
		return y;
	}
	
	public boolean isAlwaysOnTop() {
		return alwaysOnTop;
	}
	
	public boolean isAuthenticate() {
		return authenticate;
	}
	
	public boolean isBlendKeepManifest() {
		return blendKeepManifest;
	}
	
	public boolean isDemo() {
		return demo;
	}
	
	public boolean isExitRequested() {
		return exitRequested;
	}
	
	public boolean isFullscreen() {
		return fullscreen;
	}
	
	public boolean isKeepUsername() {
		return keepUsername;
	}
	
	public boolean isMaximized() {
		return maximized;
	}
	
	public boolean isNoExit() {
		return noExit;
	}
	
	public boolean isNoFrame() {
		return noFrame;
	}
	
	public boolean isPrintDump() {
		return printDump;
	}
	
	public boolean isPrintHelp() {
		return printHelp;
	}
	
	public boolean isPrintVersion() {
		return printVersion;
	}
	
	public boolean isSaveLastLogin() {
		return saveLastLogin;
	}
	
	public boolean isUseLastLogin() {
		return useLastLogin;
	}
	
	public void setExitRequested(boolean exitRequested) {
		this.exitRequested = exitRequested;
	}
	
	public void setJar(String jar) {
		this.jar = jar;
	}
	
	public void setOptionsFileFrom(String optionsFileFrom) {
		this.optionsFileFrom = optionsFileFrom;
	}
	
	public void setParentDir(String parentDir) {
		this.parentDir = parentDir;
	}
	
	public void setPassword(String password) {
		this.password = password;
	}
	
	public void setSessionId(String sessionId) {
		this.sessionId = sessionId;
	}
	
	public void setUsername(String username) {
		this.username = username;
	}
	
	@Override
	public String toString() {
		StringBuilder value = new StringBuilder();
		value.append("jarDir (exists: ").append(new File(jarDir).exists()).append("): ").append(jarDir).append("\n");
		value.append("jar (exists: ").append(new File(jar).exists()).append("): ").append(jar).append("\n");
		value.append("lwjglDir (exists: ").append(new File(lwjglDir).exists()).append("): ").append(lwjglDir).append("\n");
		value.append("password: ").append(password).append("\n");
		value.append("nativeDir (exists: ").append(new File(nativeDir).exists()).append("): ").append(nativeDir).append("\n");
		value.append("additionalJars:").append("\n");
		for (String additionalJar : additionalJars) {
			value.append("    ").append(additionalJar).append("\n");
		}
		value.append("noFrame: ").append(noFrame).append("\n");
		value.append("optionsFileFrom (exists: ").append(new File(optionsFileFrom).exists()).append("): ").append(optionsFileFrom).append("\n");
		value.append("options:").append("\n");
		for (String option : options) {
			value.append("    ").append(option).append("\n");
		}
		value.append("demo: ").append(demo).append("\n");
		value.append("parentDir (exists: ").append(new File(parentDir).exists()).append("): ").append(parentDir).append("\n");
		value.append("applet: ").append(appletToLoad).append("\n");
		value.append("blendWith: ").append("\n");
		for (String file : blendWith) {
			value.append("	(exists: ").append(new File(file).exists()).append("): ").append(file).append("\n");
		}
		value.append("blendJarName: ").append(blendJarName).append("\n");
		value.append("blendKeepManifest: ").append(blendKeepManifest).append("\n");
		value.append("port: ").append(port).append("\n");
		value.append("server: ").append(server).append("\n");
		value.append("authenticate: ").append(authenticate).append("\n");
		value.append("authenticationFailureBehavior: ").append(authenticationFailureBehavior).append("\n");
		value.append("keepAliveTick: ").append(keepAliveTick).append("\n");
		value.append("launcherVersion: ").append(launcherVersion).append("\n");
		value.append("authenticationAddress: ").append(authenticationAddress).append("\n");
		value.append("username: ").append(username).append("\n");
		value.append("useLastLogin: ").append(useLastLogin).append("\n");
		value.append("saveLastLogin: ").append(saveLastLogin).append("\n");
		value.append("keepUsername: ").append(keepUsername).append("\n");
		value.append("texturepack: ").append(texturepack).append("\n");
		value.append("maximized: ").append(maximized).append("\n");
		value.append("width: ").append(width).append("\n");
		value.append("height: ").append(height).append("\n");
		value.append("x: ").append(x).append("\n");
		value.append("y: ").append(y).append("\n");
		value.append("title: ").append(title).append("\n");
		value.append("alwaysOnTop: ").append(alwaysOnTop).append("\n");
		value.append("fullscreen: ").append(fullscreen).append("\n");
		value.append("opacity: ").append(opacity).append("\n");
		return value.toString();
	}
	
	/**
	 * Extends the additional jars with the given one, but only if there isn't
	 * already a jar added with the same filename.
	 * 
	 * @param jarToAdd the jar to add.
	 */
	private void extendAdditionalJars(String jarToAdd) {
		String jarToAddFilename = new File(jarToAdd).getName();
		
		for (String jar : additionalJars) {
			if (jar.endsWith(jarToAddFilename)) {
				return;
			}
		}
		
		additionalJars.add(jarToAdd);
	}
}
